File: app/src/main/java/com/absinthe/anywhere_/viewbuilder/entity/DynamicParamsDialogBuilder.java
package com.absinthe.anywhere_.viewbuilder.entity;

import android.content.Context;
import android.widget.EditText;
import android.widget.LinearLayout;

import com.absinthe.anywhere_.R;
import com.absinthe.anywhere_.utils.manager.Logger;
import com.absinthe.anywhere_.viewbuilder.ViewBuilder;

import java.util.ArrayList;
import java.util.List;

public class DynamicParamsDialogBuilder extends ViewBuilder {

    private List<EditText> editTextList = new ArrayList<>();
    private String[] params;

    public DynamicParamsDialogBuilder(Context context) {
        super(context);
    }

    @Override
    public void init() {
        root = new LinearLayout(mContext);
        root.setLayoutParams(Params.LL.MATCH_WRAP);
        int padding = (int) mResources.getDimension(R.dimen.bsd_edit_text_margin_horizontal);
        root.setPadding(padding, 0, padding, 0);
        ((LinearLayout) root).setOrientation(LinearLayout.VERTICAL);
    }

    public void setParams(String paramString) {
        Logger.d(paramString);
        params = paramString.split("&");

        for (String para : params) {
            Logger.d(para);
            EditText editText = new EditText(mContext);
            editText.setLayoutParams(Params.LL.MATCH_WRAP);
            editText.setSingleLine(true);
            editText.setHint(para);
            addView(editText);
            editTextList.add(editText);
        }
    }

    public String getInputParams() {
        StringBuilder sb = new StringBuilder();
        sb.append("?");
        for (int iter = 0, len = params.length; iter < len; iter++) {
            sb.append(params[iter])
                    .append("=")
                    .append(editTextList.get(iter).getText() == null
                            ? ""
                            : editTextList.get(iter).getText().toString());
            if (iter != len - 1) {
                sb.append("&");
            }
        }
        Logger.d(sb);
        return sb.toString();
    }
}

File: app/src/main/java/com/absinthe/anywhere_/viewbuilder/entity/DynamicParamsDialogBuilder.kt
package com.absinthe.anywhere_.viewbuilder.entity

import android.content.Context
import android.view.inputmethod.InputMethodManager
import android.widget.EditText
import android.widget.LinearLayout
import com.absinthe.anywhere_.R
import com.absinthe.anywhere_.utils.doOnMainThreadIdle
import com.absinthe.anywhere_.viewbuilder.ViewBuilder
import timber.log.Timber

class DynamicParamsDialogBuilder(context: Context) : ViewBuilder(context) {

  private val paramsMap: MutableMap<String, EditText> = mutableMapOf()

  init {
    root = LinearLayout(context).apply {
      layoutParams = LinearLayout.LayoutParams(
        LinearLayout.LayoutParams.MATCH_PARENT,
        LinearLayout.LayoutParams.WRAP_CONTENT
      )

      val padding = context.resources.getDimension(R.dimen.bsd_edit_text_margin_horizontal).toInt()
      setPadding(padding, 0, padding, 0)
      orientation = LinearLayout.VERTICAL
    }
  }

  fun setParams(paramString: String) {
    Timber.d(paramString)
    val params = paramString.split("&").toTypedArray()

    for (para in params) {
      val editText = EditText(mContext).apply {
        layoutParams = LinearLayout.LayoutParams(
          LinearLayout.LayoutParams.MATCH_PARENT,
          LinearLayout.LayoutParams.WRAP_CONTENT
        )
        isSingleLine = true
        hint = para
      }

      addView(editText)
      paramsMap[para] = editText
    }
    paramsMap[params[0]]?.apply {
      requestFocus()

      doOnMainThreadIdle({
        val inputManager =
          context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        inputManager.showSoftInput(this, 0)
      })
    }

  }

  val inputParams: String
    get() {
      val sb = StringBuilder().apply {
        append("?")

        for (iterator in paramsMap) {
          append("${iterator.key}=${iterator.value.text}&")
        }

        removeSuffix("&")
      }

      Timber.d(sb.toString())
      return sb.toString()
    }
}
